streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreSegmentValueFormatter.java

```diff
@@ -0,0 +1,554 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Helper utility for managing the bytes layout of the value stored in segments of the {@link RocksDBVersionedStore}.
+ * The value format is:
+ * <pre>
+ *     <next_timestamp> + <min_timestamp> + <list of <timestamp, value_size>, reverse-sorted by timestamp> + <list of values, forward-sorted by timestamp>
+ * </pre>
+ * where:
+ * <ul>
+ * <li>{@code next_timestamp} is the validTo timestamp of the latest record version stored in this
+ * segment,</li>
+ * <li>{@code min_timestamp} is the validFrom timestamp of the earliest record version stored
+ * in this segment, and</li>
+ * <li>Negative {@code value_size} is used to indicate that the value stored is a tombstone,
+ * in order to distinguish from empty array which has {@code value_size} of zero. In practice,
+ * {@code value_size} is always set to -1 for the tombstone case, though this need not be true
+ * in general.</li>
+ * </ul>
+ * <p>
+ * Note that the value format above does not store the number of record versions contained in the
+ * segment. It is not necessary to store this information separately because this information is
+ * never required on its own. Record versions are always deserialized in order, and we can
+ * determine when we have reached the end of the list based on whether the (validFrom) timestamp of
+ * the record version equals the {@code min_timestamp}.
+ * <p>
+ * Additionally, there is one edge case / exception to the segment value format described above.
+ * If the latest record version (for a particular key) is a tombstone, and the segment in which
```

Comment:
I tried to clarify what's going on in https://github.com/apache/kafka/pull/13186.

It's not only if there's nothing to be deleted from the latest value store. Even if there's nothing in the latest value store because the latest value is a tombstone, as long as that tombstone is in the same segment as the new one being inserted then we also don't hit the edge case. We really only hit the edge case if the store is empty (or empty from the perspective of the put operation, i.e., whatever data is contained in the store is only present in older segments which are not relevant for the specific put operation). I spliced together your explanation and mine. Thanks for the contribution :)